### PR TITLE
[move core modules] Option module

### PR DESCRIPTION
**Pull request by sblackshear**
RFC because I want to get some consensus on the API before adding too many tests.

Option is a very useful abstraction. We have long discussed adding this to Move. I am starting to see code in the stdlib that would really benefit from Option, so now seems like a good time. Here's what this PR does:

- Introduces the `Option` module, implemented as a wrapper around a vector of size 0 or 1.
- It can be constructed as empty (none) or nonempty (some)
- Has all of the utility functions you would expect for reading the option
- On the mutation front, an option can be extracted (take the value out, turning it into None), filled (insert a value into it), unpacked (destroy the Some option and hand you the value), or destroyed (wipe out a None option)
- I modified `ValidatorConfig2` to demonstrate where `Option` would lead to simpler code.
- I think there are several other places where we might benefit from Option. In particular, the pattern of an `XCapability { is_certified: false }` that can be published by anyone + certified by an authority might benefit.

Note: we have previously considered a native option type (e.g., to enable safe vector APIs that return options). The presence of this pure Move option doesn't rule out that path.

## Motivation

See above

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Y

## Test Plan

New functional tests that exercise the new APIs

### CODE REVIEW

File: language/stdlib/modules/option.move [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            Vector::borrow(vec_ref, 0)
+        } else {
+            default_ref
+        }
+    }
+
+    // TODO: not clear we want both this and borrow_default
+    // Return the value inside `t` if it holds one
+    // Return `default` if `t` does not hold a value
+    public fun get_default<Element: copyable>(t: &T<Element>, default: Element): Element {
```
Comment (tnowacki) [inline language/stdlib/modules/option.move:55]: I don't think you need this, since if you have `Option:: get_default(opt, e)`  you can always rewrite it as `*Option:: borrow_default(opt, &e)`

File: language/stdlib/modules/option.move [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        if (Vector::is_empty(vec_ref)) {
+            *Vector::borrow(vec_ref, 0)
+        } else {
+            default
+        }
+    }
+
+    // Convert the none option `t` to a some option by adding `e`.
+    // Aborts if `t` already holds a value
+    public fun fill<Element>(t: &mut T<Element>, e: Element) {
```
Comment (tnowacki) [inline language/stdlib/modules/option.move:66]: ~~Feels like a weird API function to have. What's the use case you see?~~

Nevermind, I think this is useful because of resources

Normally, you would just assign the option
```
*o = Option::some(e)
```
but you cant do that if `Element` is a resource
You need
```
Option::fill(o, e)
```

File: language/stdlib/modules/option.move (hunk 1 above)
Comment (tnowacki) [reply language/stdlib/modules/option.move:55]: I think this API would be nice if somehow it took an owned version of `Option`? Then at least it feels like it can help in situations where you don't want to copy?

So
```
    public fun get_default<Element: copyable>(t: T<Element>, default: Element): Element {
```

File: language/stdlib/modules/option.move [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    // ===== APIs that might be controversial + I am inclined to exclude ====
+
+    // Why controversial: do we want interior mutability for an option `o` given that you could
+    // just (1) reasign `o` if it's copyable or (2) extract, then fill?
+    //
+    // Return an immutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow_mut<Element>(t: &mut T<Element>): &Element {
```
Comment (tnowacki) [inline language/stdlib/modules/option.move:104]: Probably want `&mut Element` at the end :)

But I think this is useful

File: language/stdlib/modules/option.move [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        Vector::borrow_mut(&mut t.vec, 0)
+    }
+
+    // Why controversial: same as above
+    //
+    // Swap the old value inside `t` with `e` and return the old value
+    // Aborts if `t` does not hold a value
+    public fun swap<Element>(t: &mut T<Element>, e: Element): Element {
```
Comment (tnowacki) [inline language/stdlib/modules/option.move:112]: This seems like a must have with resources

File: language/stdlib/modules/option.move [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    // Aborts if `t` holds a value
+    public fun destroy_none<Element>(t: T<Element>) {
+        let T { vec } = t;
+        Vector::destroy_empty(vec)
+    }
+
+    // ===== APIs that might be controversial + I am inclined to exclude ====
+
+    // Why controversial: do we want interior mutability for an option `o` given that you could
+    // just (1) reasign `o` if it's copyable or (2) extract, then fill?
```
Comment (tnowacki) [inline language/stdlib/modules/option.move:100]: I think you want exposed mutability of the element 
Imagine something like 
```
o: &mut Option::T<Coin::T>;
...
Coin::withdraw(Option::borrow_mut(o), 10)
```
Without borrow mut that becomes
```
let coin = Option::extract(o);
Coin::withdraw(&mut coin, 10);
Option::fill(o, coin)
```

File: language/stdlib/modules/option.move [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+    // Return true if `t` holds a value
+    public fun is_some<Element>(t: &T<Element>): bool {
+        !Vector::is_empty(&t.vec)
+    }
+
+    // Return true if the value in `t` is equal to `e_ref`
+    // Always returns `false` if `t` does not hold a value
+    public fun contains<Element>(t: &T<Element>, e_ref: &Element): bool {
+        Vector::borrow(&t.vec, 0) == e_ref
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/option.move:31]: Oops, good catch.

File: language/stdlib/modules/option.move [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    public fun borrow<Element>(t: &T<Element>): &Element {
+        Vector::borrow(&t.vec, 0)
+    }
+
+    // TODO: not clear this API is a good idea
+    // Return a reference to the value inside `t` if it holds one
+    // Return `default_ref` if `t` does not hold a value
+    public fun borrow_default<Element>(t: &T<Element>, default_ref: &Element): &Element {
+        let vec_ref = &t.vec;
+        if (Vector::is_empty(vec_ref)) {
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/option.move:45]: On `is_some` vs using the vector functions directly: I was trying to avoid the overhead of an extra call. In the long run, maybe we can do something like allowing you to write `is_some` (which is definitely clearer) and giving the compiler a hint to inline or letting it decide on its own...

File: language/stdlib/modules/option.move [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+            Vector::borrow(vec_ref, 0)
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/option.move:46]: Same as above re avoiding extra calls (and a few below too...)

File: language/stdlib/modules/option.move [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    // Return an immutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow<Element>(t: &T<Element>): &Element {
+        Vector::borrow(&t.vec, 0)
+    }
+
+    // TODO: not clear this API is a good idea
+    // Return a reference to the value inside `t` if it holds one
+    // Return `default_ref` if `t` does not hold a value
+    public fun borrow_default<Element>(t: &T<Element>, default_ref: &Element): &Element {
```
Comment (tzakian) [inline language/stdlib/modules/option.move:43]: I think having this API makes sense in the presence of resources. Additionally, this is a much more general API that has as a special case `get_default` -- the user can decide if they need to/can copy into an owned value after `borrow_default`ing based upon their local kind restrictions on the type held within the `Option`. For these reasons, I would say that we should keep this and remove `get_default`. 

As an examination of this in more depth:

If `T` were `Any` or `Resource` (e.g. if we had the following function declaration: `fun x<T>(t: &Option<T>, dflt: T, ....)`):
* If we kept `get_default` but removed `borrow_default` we would be forced within this function to unpack `t` as follows:
```rust
let x = if (Option::is_none(t)) {
   &dflt
} else {
   Option::borrow(t)
};
```
* If we kept `borrow_default` but removed `get_default`:
```rust
let x = Option::borrow_default(t, &dflt);
```

In the case where `T` is `copyable` (`fun x<T: copyable>(t: &Option<T>, dflt: T, ...)`) we have:
* If we got rid of `get_default` but keep `borrow_default`:
```rust
let t = *Option::borrow_default(t, &dflt);
```
* If we got rid of `borrow_default` but kept `get_default`:
```rust
let t = Option::get_default(t, dflt);
```

File: language/stdlib/modules/option.move (hunk 5 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move:100]: Yeah, fair point. The case where the callee wants to mutate the thing inside of the option is a good motivation for this case. Will add this

File: language/stdlib/modules/option.move (hunk 1 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move:55]: > always rewrite it as `*Option:: borrow_default(opt, &e)`

I think it's fair that this isn't so much worse to write, but it does add an extra local, assignment, borrow, and dereference under the hood that can't be optimized away...

File: language/stdlib/modules/option.move (hunk 9 above)
[the remark this answers is not in the document]
Comment (tzakian) [reply language/stdlib/modules/option.move:43]: I feel as though the `or` or `or_else` type of chaining only really works syntactically with `.` syntax, and in Move syntax would feel somewhat cumbersome. I would much prefer to rename this maybe to something like `borrow_with_default`. What do you think?

[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/option.move:43]: I like the sound of `with_default`.

File: language/stdlib/modules/option.move (hunk 7 above)
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/stdlib/modules/option.move]: I feel like avoiding proper encapsulation for an extra call feels like a bad idea. I wouldn't worry too much about the fine grained gas costings of these functions (as we don't have custom code publishing). To me, we should prioritize readability and correctness over gas cost. Which I think would mean developing a small set of "core" functions that deal with the vector field directly, everything else is defined in terms of those functions. 

In short, I really think you should use `is_some` here. I don't think the brittleness is worth the gas cost.

File: language/stdlib/modules/option.move (hunk 1 above)
Comment (tnowacki) [reply language/stdlib/modules/option.move]: I'm not convinced those 4 extra operations are a big deal for that specific case. We cannot write library functions that are the most optimal for every possible case. We have to pick ones that the most convenient for the common cases. Thus, I think that `get_default` with a non-reference first value feels better for this operation. 

The copying reference case can be handled with the borrow+dereference snippet

File: language/stdlib/modules/option.move [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    // Aborts if `t` already holds a value
+    public fun fill<Element>(t: &mut T<Element>, e: Element) {
+        let vec_ref = &mut t.vec;
+        if (Vector::is_empty(vec_ref)) Vector::push_back(vec_ref, e)
+        else abort(99)
+    }
+
+    // Convert a `some` option to a `none` by removing and returning the value stored inside `t`
+    // Aborts if `t` does not hold a value
+    public fun extract<Element>(t: &mut T<Element>): Element {
```
Comment (dariorussi) [inline language/stdlib/modules/option.move:67]: why not taking the `T<Element>` instead of a reference? Essentially the destroy operation. 
And maybe without some of the helpers below...

File: language/stdlib/modules/option.move [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+    // Return a mutable reference to the value inside `t`
+    // Aborts if `t` does not hold a value
+    public fun borrow_mut<Element>(t: &mut T<Element>): &mut Element {
+        Vector::borrow_mut(&mut t.vec, 0)
+    }
+
+    // Swap the old value inside `t` with `e` and return the old value
+    // Aborts if `t` does not hold a value
+    public fun swap<Element>(t: &mut T<Element>, e: Element): Element {
```
Comment (dariorussi) [inline language/stdlib/modules/option.move:79]: this would be equivalent to someone calling `extract` and then `some` right?

File: language/stdlib/modules/option.move [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    public fun destroy_some<Element>(t: T<Element>): Element {
+        let T { vec } = t;
+        let elem = Vector::pop_back(&mut vec);
+        Vector::destroy_empty(vec);
+        elem
+    }
+
+    // Unpack `t`
+    // Aborts if `t` holds a value
+    public fun destroy_none<Element>(t: T<Element>) {
```
Comment (dariorussi) [inline language/stdlib/modules/option.move:104]: is this needed when `t` is a resource?

File: language/stdlib/modules/option.move [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+    // Aborts if `t` does not hold a value
+    public fun swap<Element>(t: &mut T<Element>, e: Element): Element {
+        let vec_ref = &mut t.vec;
+        let old_value = Vector::pop_back(vec_ref);
+        Vector::push_back(vec_ref, e);
+        old_value
+    }
+
+    // Destroys `t.` If `t` holds a value, return it. Returns `default` otherwise
+    public fun destroy_with_default<Element: copyable>(t: T<Element>, default: Element): Element {
```
Comment (dariorussi) [inline language/stdlib/modules/option.move:87]: some of those look like helpers around optimizations path. Thoughts?

File: language/stdlib/modules/option.move (hunk 11 above)
Comment (tnowacki) [reply language/stdlib/modules/option.move:79]: You could do it with `fill` if we keep that.
`let old = Option::extract(o); Option::fill(o, new); old` == `Option::swap(o, new)`

I personally really like `swap` for the resource case. I think the `extract` + `fill` option feels a bit too imperative for me

File: language/stdlib/modules/option.move (hunk 13 above)
Comment (tnowacki) [reply language/stdlib/modules/option.move:87]: I agree that the current setup definitely feels overtuned for optimizations. 
That being said its definitely a hard line to walk between optimizations, ease of use, and minimal composability. 

Originally, I like this function because it seemed like a good corollary with `get_with_default`. Looking at it again, I feel less convinced, And I could probably live without this function. 

To some degree, I almost feel that the APIs should focus just on minimal composability + ease of use. Optimizations almost feel like something that would want to be done on a one off basis using the minimal components defined in the module. I can go back and forth on this a lot, but it's how I'm feeling right now :)

File: language/stdlib/modules/option.move (hunk 10 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move:67]: I think you need this one when the `Option` is a field of another struct/resource. For example, the classic `Escrow<T1, T2> { goods1: Option<T1>, goods2: Option<T2> }`) would use this API to extract the goods (since each party needs to do this separately).

File: language/stdlib/modules/option.move (hunk 11 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move:79]: Agreed that it's equivalent to `extract` + `fill`. I think there are compelling arguments for both of those (c.f. `Escrow`), so this isn't strictly required. I don't mind having it, though.

File: language/stdlib/modules/option.move (hunk 13 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move:87]: I could imagine this being useful in the case where you're `Unpack`ing a struct with a copyable option inside. But this is probably the first function I would kill if forced to eliminate one.

File: language/stdlib/modules/option.move (hunk 12 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move:104]: Right; either when `t` is a resource or you're writing generic code that wraps this (e.g., `OptionWrapper<T> { t: Option<T> }` would need this when unpacking `OptionWrapper`).

File: language/stdlib/modules/option.move (hunk 9 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move]: I agree with all of this except for the conclusion that we should eliminate `get_default` For the `let t = *Option::borrow_default(t, &dflt);` alternative, consider the (common) case where `dflt` is a literal. The source language will still let you write something like `&8`, but under the hood it's introducing a new local variable, a `StLoc` instruction to populate it, a `BorrowLoc` to pass it to the caller, then a `ReadRef` to match the `*`. That feels like a lot of extra work that the caller might prefer not to do (even if the syntax makes it easy...).

File: language/stdlib/modules/option.move (hunk 1 above)
Comment (sblackshear) [reply language/stdlib/modules/option.move]: > We cannot write library functions that are the most optimal for every possible case. We have to pick ones that the most convenient for the common cases

I agree. I think this function is both convenient and optimal for a common case, however. As weak evidence: the use of `Option` in `ValidatorConfig2` uses this API.

File: language/stdlib/modules/option.move (hunk 7 above)
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/stdlib/modules/option.move]: I agree with the general design principle or writing a set of core functions that manipulate the richer API (e.g., vector) and  then writing everything in terms of those. But: 
- I don't think it makes sense to write a core function for `Vector::borrow` (since it would be a wrapper that does exactly the same thing)
- I personally find it less readable to mix locally defined functions like `is_some` with raw accesses to `Vector` than to use all raw accesses

I'm happy to change this if it's a sticking point  or if I'm missing some cleaner way to write this in terms of core functions with no `Vector` accesses at all.

**Comment by sblackshear** [comment]
Added tests, so no longer RFC!

**Comment by sblackshear** [comment]
Any other blockers on this @tnowacki @tzakian @dariorussi?
